buildconfig/ci/travis/.travis_osx_before_install.sh

```diff
@@ -83,15 +83,105 @@ function retry {
 }
 
 function install_or_upgrade {
-	set +e
-    if brew ls --versions "$1" >/dev/null; then
-        echo "package already installed: $@"
+  local deps=""
+  if (brew info "$1" | grep "(bottled)" >/dev/null); then
+    deps=$(brew deps "$1")
+  else
+    deps=$(brew deps --include-build "$1")
+  fi
+  if [[ "$deps" ]]; then
+    echo -n "$1 dependencies: "
+    echo $deps
+    while read -r dependency; do
+      echo "$1: Install dependency $dependency."
+      install_or_upgrade "$dependency"
+    done <<< "$deps"
+  fi
+
+  if (brew ls --versions "$1" >/dev/null) && ! (brew outdated | grep "$1" >/dev/null); then
+    echo "$1 is already installed and up to date."
+  else
+    if (brew outdated | grep "$1" >/dev/null); then
+      echo "$1 is installed but outdated."
+      if (brew info "$1" | grep "(bottled)" >/dev/null); then
+        echo "$1: Found bottle."
+        retry brew upgrade "$1"
+        return 0
+      else
+        brew uninstall --ignore-dependencies "$1"
```

Comment:
I'm unsure about this especially. Can't you upgrade and create a bottle at the same time?